File: src/policy.ts
export const MAX_FILE_BYTES = 256 * 1024;
export const MAX_DIRECTORY_DEPTH = 4;
export const MAX_DIRECTORY_ENTRIES = 250;

const SECRET_KEY_PATTERN = /(api[_-]?key|token|secret|password|passwd|credential|private[_-]?key|auth|bearer)/i;

export function isSecretLikeKey(key: string): boolean {
  return SECRET_KEY_PATTERN.test(key);
}

export function captureStatusForKey(key: string): "redacted" | "omitted" {
  return isSecretLikeKey(key) ? "redacted" : "omitted";
}

export function redactStructuredValue(value: unknown): unknown {
  if (Array.isArray(value)) {
    return value.map((entry) => redactStructuredValue(entry));
  }

  if (!value || typeof value !== "object") {
    return value;
  }

  const redacted: Record<string, unknown> = {};
  for (const [key, nestedValue] of Object.entries(value)) {
    if (isSecretLikeKey(key)) {
      redacted[key] = "[redacted]";
    } else if (key === "env" && nestedValue && typeof nestedValue === "object" && !Array.isArray(nestedValue)) {
      redacted.envKeys = Object.keys(nestedValue);
    } else {
      redacted[key] = redactStructuredValue(nestedValue);
    }
  }

  return redacted;
}

export function ignoredDirectory(name: string): boolean {
  return new Set([
    ".git",
    "node_modules",
    "dist",
    "build",
    ".cache",
    "cache",
    "caches",
    "logs",
    "log",
    ".next",
    "coverage",
    ".turbo"
  ]).has(name);
}

File: internal/gandalfcore/policy/policy.go
package policy

import (
	"encoding/json"
	"regexp"

	"github.com/qyinm/gandalf/internal/gandalfcore/types"
)

const (
	MaxFileBytes        = 256 * 1024
	MaxDirectoryDepth   = 4
	MaxDirectoryEntries = 250
)

var secretKeyPattern = regexp.MustCompile(`(?i)(api[_-]?key|token|secret|password|passwd|credential|private[_-]?key|auth|bearer)`)

func RestorePolicyFor(kind types.EvidenceKind) types.RestorePolicy {
	switch kind {
	case types.KindAgentInstruction, types.KindAgentConfig, types.KindSkill, types.KindExtension:
		return types.RestoreFullContent
	case types.KindMcpServer, types.KindPermission, types.KindHook:
		return types.RestoreStructuredFields
	case types.KindEnvKey:
		return types.RestoreKeyInventory
	default:
		return types.RestoreNotSupported
	}
}

func IsSecretLikeKey(key string) bool {
	return secretKeyPattern.MatchString(key)
}

func CaptureStatusForKey(key string) string {
	if IsSecretLikeKey(key) {
		return "redacted"
	}
	return "omitted"
}

func RedactStructuredValue(value json.RawMessage) (json.RawMessage, error) {
	var decoded any
	if err := json.Unmarshal(value, &decoded); err != nil {
		return value, err
	}
	redacted := redactValue(decoded)
	out, err := json.Marshal(redacted)
	if err != nil {
		return nil, err
	}
	return out, nil
}

func redactValue(value any) any {
	switch v := value.(type) {
	case map[string]any:
		return redactObject(v)
	case []any:
		out := make([]any, len(v))
		for i, item := range v {
			out[i] = redactValue(item)
		}
		return out
	default:
		return value
	}
}

func redactObject(obj map[string]any) map[string]any {
	redacted := make(map[string]any, len(obj))
	for key, nested := range obj {
		if IsSecretLikeKey(key) {
			redacted[key] = "[redacted]"
			continue
		}
		if key == "env" {
			if envMap, ok := nested.(map[string]any); ok {
				keys := make([]string, 0, len(envMap))
				for k := range envMap {
					keys = append(keys, k)
				}
				redacted["envKeys"] = keys
				continue
			}
		}
		redacted[key] = redactValue(nested)
	}
	return redacted
}

func IgnoredDirectory(name string) bool {
	switch name {
	case ".git", "node_modules", "dist", "build", ".cache", "cache", "caches", "logs", "log", ".next", "coverage", ".turbo":
		return true
	default:
		return false
	}
}
